topology2: add sof-mtl-cs42l42-l0-max98363-l2-4ch support

yongzhi1 opened the pull request:
cs42l42 headset codec on link0 & 2xmax98363 on link 2.

plbossart commented on tools/topology/topology2/sof-ace-tplg/tplg-targets.cmake:
```diff
@@ -20,6 +20,10 @@ DEEPBUFFER_FW_DMA_MS=100"
 "cavs-sdw\;sof-mtl-rt711-4ch\;PLATFORM=mtl,NUM_DMICS=4,DMIC0_ID=2,DMIC1_ID=3,NUM_HDMIS=0,\
 PREPROCESS_PLUGINS=nhlt,NHLT_BIN=nhlt-sof-mtl-rt711-4ch.bin,DEEPBUFFER_FW_DMA_MS=100"
 
+"cavs-sdw\;sof-mtl-cs42l42-l0-max98363-l2-4ch\;PLATFORM=mtl,\
```
how would the topology deal with the amps on link2, there's nothing that tells us to use link2 anywhere in the macros.

yongzhi1 replied:
Ack, think spk amp is on link2 by default now, maybe we should use macro to config it explicitly for future-proof: 
https://github.com/thesofproject/sof/blob/main/tools/topology/topology2/platform/intel/sdw-amp-generic.conf

bardliao replied:
> I think this topology is just confusing. I have no idea what the BE_ID refers to, if this is the index of the dailink or the physical link. @bardliao any comments?

@plbossart @yongzhi1 BE_ID refers to dailink index. SDW_SPK_STREAM refers to the physical link. Default string 'SDW1-Playback' is link 1. So, you probably need `SDW_SPK_STREAM=SDW2-Playback'`

yongzhi1 replied:
Ack, will update SDW_SPK_STREAM, but is this just a token referenced internally? fyi here is the config diagram from
from https://review.coreboot.org/c/coreboot/+/73280:
![sdw](https://user-images.githubusercontent.com/19287706/228912950-2c260697-5356-4a9a-9e9d-8f3b89cd91b7.gif)

bardliao replied:
@yongzhi1 SDW_SPK_STREAM is not only referenced internally. It is also the name of Object.Dai.ALH and should match to the dai link name. For some reason, SDW_SPK_STREAM is referenced to copier's stream name, too. However, copier's stream name doesn't have to be the same as DAI name.

yongzhi1 commented:
V2 update:

- Add SDW_SPK_STREAM and SDW_SPK_IN_STREAM override to be matched by CPU DAI number in sof_sdw.c.

bardliao commented on tools/topology/topology2/sof-ace-tplg/tplg-targets.cmake:
```diff
@@ -20,6 +20,12 @@ DEEPBUFFER_FW_DMA_MS=100"
+# SDW + DMIC + HDMI with SDW_AMP_FEEDBACK
+"cavs-sdw\;sof-mtl-cs42l42-l0-max98363-l2-4ch\;PLATFORM=mtl,\
+NUM_SDW_AMPS=2,SDW_SPK_STREAM=SDW2-Playback,SDW_SPK_IN_STREAM=SDW2-Capture,\
+NUM_DMICS=4,DMIC0_ID=2,DMIC1_ID=3,NUM_HDMIS=4,DEEPBUFFER_FW_DMA_MS=100,\
```
You need `PDM1_MIC_A_ENABLE=1,PDM1_MIC_B_ENABLE=1` as `NUM_DMICS=4`

yongzhi1 replied:
Ack, will update in next version.

sys-pt1s commented:
Can one of the admins verify this patch?

yongzhi1 commented:
V3 update:
1.  Add  PDM1_MIC_A_ENABLE=1,PDM1_MIC_B_ENABLE=1. 
2.  Changed NUM_SDW_AMPS=1.

yongzhi1 commented:
V4 update:
1. Add SDW_AMP_FEEDBACK=false since MX98363 does not have feedback path.
2. Use their default value for DMIC0_ID/DMIC1_ID.

plbossart commented on tools/topology/topology2/sof-ace-tplg/tplg-targets.cmake:
```diff
@@ -20,6 +20,13 @@ USE_CHAIN_DMA=true,DEEPBUFFER_FW_DMA_MS=100"
 
+# SDW + DMIC + HDMI
+"cavs-sdw\;sof-mtl-cs42l42-l0-max98363-l2-4ch\;PLATFORM=mtl,\
+NUM_SDW_AMPS=1,SDW_SPK_STREAM=SDW2-Playback,SDW_SPK_IN_STREAM=SDW2-Capture,SDW_AMP_FEEDBACK=false,\
```
is this true? isn't this codec capable of providing feedback?

yongzhi1 replied:
From the datasheet, the codec does not have feedback capability.

plbossart commented on tools/topology/topology2/sof-ace-tplg/tplg-targets.cmake:
```diff
@@ -20,6 +20,13 @@ USE_CHAIN_DMA=true,DEEPBUFFER_FW_DMA_MS=100"
+"cavs-sdw\;sof-mtl-cs42l42-l0-max98363-l2-4ch\;PLATFORM=mtl,\
+NUM_SDW_AMPS=1,SDW_SPK_STREAM=SDW2-Playback,SDW_SPK_IN_STREAM=SDW2-Capture,SDW_AMP_FEEDBACK=false,\
+NUM_DMICS=4,PDM1_MIC_A_ENABLE=1,PDM1_MIC_B_ENABLE=1,\
+NUM_HDMIS=4,DEEPBUFFER_FW_DMA_MS=100,\
```
HDMI=3, we never use 4 in practice so let's keep things simple(r).

yongzhi1 replied:
Ack, also noticed NUM_SDW_AMPS was changed to NUM_SDW_AMP_LINKS, will update in next version.

yongzhi1 commented:
v5 update:
- HDMI use default setting found in cavs_sdw.conf(NUM_HDMIS etc.)
- Changed NUM_SDW_AMPS to NUM_SDW_AMP_LINKS
- Updated DMIC pcm id to 99

yongzhi1 commented:
Hopeful this is the final version:

Added SDW_JACK_CAPTURE_CH=1.

yongzhi1 commented on tools/topology/topology2/sof-ace-tplg/tplg-targets.cmake:
```diff
@@ -45,4 +45,11 @@ NHLT_BIN=nhlt-sof-mtl-max98357a-rt5682.bin,DEEPBUFFER_FW_DMA_MS=10,HEADSET_SSP_D
+"cavs-sdw\;sof-mtl-sdw-cs42l42-l0-max98363-l2\;PLATFORM=mtl,NUM_DMICS=4,PDM1_MIC_A_ENABLE=1,\
+PDM1_MIC_B_ENABLE=1,DMIC0_PCM_ID=99,DMIC0_ID=3,DMIC1_ID=4,USE_CHAIN_DMA=true,\
+PREPROCESS_PLUGINS=nhlt,NHLT_BIN=nhlt-sof-mtl-sdw-cs42l42-l0-max98363-l2.bin,\
+NUM_SDW_AMP_LINKS=1,SDW_SPK_STREAM=SDW2-Playback,SDW_SPK_IN_STREAM=SDW2-Capture,SDW_AMP_FEEDBACK=false,\
```
Ack, SDW_SPK_IN_STREAM is on SDW1 by default, since we do not have feedback, the override is redundant.

yongzhi1 commented on tools/topology/topology2/sof-ace-tplg/tplg-targets.cmake:
```diff
@@ -45,4 +45,11 @@ NHLT_BIN=nhlt-sof-mtl-max98357a-rt5682.bin,DEEPBUFFER_FW_DMA_MS=10,HEADSET_SSP_D
+
+# SDW + DMIC + HDMI
+"cavs-sdw\;sof-mtl-sdw-cs42l42-l0-max98363-l2\;PLATFORM=mtl,NUM_DMICS=4,PDM1_MIC_A_ENABLE=1,\
+PDM1_MIC_B_ENABLE=1,DMIC0_PCM_ID=99,DMIC0_ID=3,DMIC1_ID=4,USE_CHAIN_DMA=true,\
```
> DMIC0_PCM_ID=99
> 
> For SoundWire devices we use PCM 10 for DMIC capture.

Ack.

yongzhi1 commented:
Update:

-  use default DMIC0_PCM_ID which is 10
-  remove SDW_SPK_IN_STREAM override

plbossart commented on tools/topology/topology2/sof-ace-tplg/tplg-targets.cmake:
```diff
@@ -45,4 +45,11 @@ NHLT_BIN=nhlt-sof-mtl-max98357a-rt5682.bin,DEEPBUFFER_FW_DMA_MS=10,HEADSET_SSP_D
+PDM1_MIC_B_ENABLE=1,DMIC0_ID=3,DMIC1_ID=4,USE_CHAIN_DMA=true,\
+PREPROCESS_PLUGINS=nhlt,NHLT_BIN=nhlt-sof-mtl-sdw-cs42l42-l0-max98363-l2.bin,\
+NUM_SDW_AMP_LINKS=1,SDW_SPK_STREAM=SDW2-Playback,SDW_AMP_FEEDBACK=false,\
+SDW_JACK_CAPTURE_CH=1,DEEPBUFFER_FW_DMA_MS=100,DEEPBUFFER_D0I3_COMPATIBLE=true"
```
@yongzhi1 FYI when you use Co-developed-by, the Signed-off-by needs to follow immediately for kernel commits. this is topology so probably ok, but I thought I'd mention it.
